source/mir/base64.d

```diff
@@ -0,0 +1,317 @@
+/++
+$(H1 @nogc Simple Base64 parsing)
+
+License: $(HTTP www.apache.org/licenses/LICENSE-2.0, Apache-2.0)
+Authors: Harrison Ford
+Copyright: 2021 Harrison Ford, Kaleidic Associates Advisory Limited, Symmetry Investments
++/
+module mir.base64;
+import mir.ndslice.topology;
+import core.bitop : bswap;
+
+package static immutable base64DecodeInvalidCharMsg = "Invalid character encountered.";
+package static immutable base64DecodeInvalidLenMsg = "Cannot decode a buffer with given length (not a multiple of 4, missing padding?)";
+version(D_Exceptions) {
+    package static immutable base64DecodeInvalidCharException = new Exception(base64DecodeInvalidCharMsg);
+    package static immutable base64DecodeInvalidLenException = new Exception(base64DecodeInvalidLenMsg);
+}
+
+// NOTE: I do not know if this would work on big-endian systems.
+// Needs further testing to figure out if it *does* work on them.
+
+// Technique borrowed from http://0x80.pl/notesen/2016-01-12-sse-base64-encoding.html#branchless-code-for-lookup-table
+private char lookup_encoding(ubyte i) @safe @nogc pure {
+    assert(i < 64);
+
+    ubyte shift;
+
+    if (i < 26)
+    {
+        // range A-Z
+        shift = 'A';
+    }
+    else if (i >= 26 && i < 52)
+    {
+        // range a-z
+        shift = 'a' - 26;
+    }
+    else if (i >= 52 && i < 62)
+    {
+        // range 0-9
+        shift = cast(ubyte)('0' - 52);
+    }
+    else if (i == 62)
+    {
+        // character plus
+        shift = cast(ubyte)('+' - 62);
+    }
+    else if (i == 63)
+    {
+        // character slash
+        shift = cast(ubyte)('/' - 63);
+    }
+
+    return cast(char)(i + shift);
+}
+
+// Do the inverse of above (convert an ASCII value into the Base64 character set)
+private ubyte lookup_decoding(char i) @safe @nogc pure
+{
+    // Branching bad, but this isn't performance sensitive
+    if (i <= 'Z' && i >= 'A') {
+        return cast(ubyte)(i - 'A');
+    }
+    else if (i <= 'z' && i >= 'a') {
+        return cast(ubyte)(i - 'a' + 26); 
+    }
+    else if (i <= '9' && i >= '0') {
+        return cast(ubyte)(i - '0' + 52);
+    }
+    else if (i == '+') {
+        return 62;
+    }
+    else if (i == '/') {
+        return 63;
+    }
+    // Just return 0 for padding,
+    // as it typically means nothing.
+    else if (i == '=') {
+        return 0;
+    }
+    else {
+        version(D_Exceptions) {
+            throw base64DecodeInvalidCharException;
+        } else {
+            assert(0, base64DecodeInvalidCharMsg);
+        }
+    }
+
+}
+
+/++
+Decode a Base64 encoded value, returning the buffer.
++/
+ubyte[] decodeBase64(scope ubyte[] buf) @safe pure
```

Comment:
`scope const(char)[] data` + appropriate changes in the other code

1. Input should be scope const
2. char should be used where input/output is text
3. don't use `buf` for input